app/src/main/java/io/gnosis/safe/ui/settings/owner/ledger/LedgerController.kt

```diff
@@ -0,0 +1,275 @@
+package io.gnosis.safe.ui.settings.owner.ledger
+
+import android.Manifest
+import android.app.Activity
+import android.bluetooth.BluetoothAdapter
+import android.bluetooth.BluetoothDevice
+import android.bluetooth.BluetoothGattCharacteristic
+import android.bluetooth.BluetoothManager
+import android.bluetooth.le.ScanCallback
+import android.bluetooth.le.ScanFilter
+import android.bluetooth.le.ScanResult
+import android.bluetooth.le.ScanSettings
+import android.content.Context
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.os.Build
+import android.os.ParcelUuid
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.Fragment
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.parseGetAddress
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.splitPath
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.unwrapADPU
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.wrapADPU
+import io.gnosis.safe.ui.settings.owner.ledger.ble.ConnectionEventListener
+import io.gnosis.safe.ui.settings.owner.ledger.ble.ConnectionManager
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.conflate
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.toHexString
+import timber.log.Timber
+import java.util.*
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.suspendCoroutine
+
+class LedgerController(val context: Context) {
+
+    var connectedDevice: BluetoothDevice? = null
+        private set
+
+    private var deviceConnectedCallback: DeviceConnectedCallback? = null
+    private var addressContinuation: Continuation<Solidity.Address>? = null
+
+    var writeCharacteristic: BluetoothGattCharacteristic? = null
+    var notifyCharacteristic: BluetoothGattCharacteristic? = null
+
+    fun loadDeviceCharacteristics() {
```

Comment:
Is this the right place in this file for a public function?

Reply:
where would be the right place?

Reply:
this fun should not be public in the first place